src/test/sql/sql_pipeline_test.cpp

```diff
@@ -0,0 +1,385 @@
+}
+
+TEST_F(SQLPipelineTest, GetParsedSQL) {
+  SQLPipeline sql_pipeline(_select_query_a);
+  const auto& parsed_sql = sql_pipeline.get_parsed_sql();
+
+  EXPECT_TRUE(parsed_sql.isValid());
+
+  auto statements = parsed_sql.getStatements();
+
+  EXPECT_EQ(statements.size(), 1ul);
+  EXPECT_EQ(statements.at(0)->type(), hsql::StatementType::kStmtSelect);
+}
+
+TEST_F(SQLPipelineTest, GetParsedSQLTwice) {
+  SQLPipeline sql_pipeline(_select_query_a);
+  sql_pipeline.get_parsed_sql();
+  const auto& parsed_sql = sql_pipeline.get_parsed_sql();
+
+  EXPECT_TRUE(parsed_sql.isValid());
+
+  auto statements = parsed_sql.getStatements();
+
+  EXPECT_EQ(statements.size(), 1ul);
+  EXPECT_EQ(statements.at(0)->type(), hsql::StatementType::kStmtSelect);
+}
+
+TEST_F(SQLPipelineTest, GetParsedSQLInvalid) {
+  SQLPipeline sql_pipeline(_invalid_sql);
+  EXPECT_THROW(sql_pipeline.get_parsed_sql(), std::exception);
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQP) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  const auto& lqp_roots = sql_pipeline.get_unoptimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_EQ(lqp_roots.at(0)->type(), LQPNodeType::Projection);
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQPTwice) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  sql_pipeline.get_unoptimized_logical_plan();
+  const auto& lqp_roots = sql_pipeline.get_unoptimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_EQ(lqp_roots.at(0)->type(), LQPNodeType::Projection);
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQPWithJoinFilter) {
+  SQLPipeline sql_pipeline(_join_query);
+
+  const auto& lqp_roots = sql_pipeline.get_unoptimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_EQ(lqp_roots.at(0)->type(), LQPNodeType::Projection);
+  EXPECT_EQ(lqp_roots.at(0)->left_child()->type(), LQPNodeType::Predicate);
+  EXPECT_EQ(lqp_roots.at(0)->left_child()->left_child()->type(), LQPNodeType::Join);
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQPInvalid) {
+  SQLPipeline sql_pipeline(_invalid_sql);
+  EXPECT_THROW(sql_pipeline.get_unoptimized_logical_plan(), std::exception);
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQPValidated) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  const auto& lqp_roots = sql_pipeline.get_unoptimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_TRUE(_contains_validate(lqp_roots.at(0)));
+}
+
+TEST_F(SQLPipelineTest, GetUnoptimizedLQPNotValidated) {
+  SQLPipeline sql_pipeline(_select_query_a, false);
+
+  const auto& lqp_roots = sql_pipeline.get_unoptimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_FALSE(_contains_validate(lqp_roots.at(0)));
+}
+
+TEST_F(SQLPipelineTest, GetOptimizedLQP) {
+  SQLPipeline sql_pipeline(_select_query_a);
+
+  const auto& lqp_roots = sql_pipeline.get_optimized_logical_plan();
+
+  EXPECT_EQ(lqp_roots.size(), 1ul);
+  EXPECT_EQ(lqp_roots.at(0)->type(), LQPNodeType::Projection);
```

Comment:
again, it's unclean to pretend to know what the optimizer does in a non-optimizer test.

Reply:
I am also not a huge fan of doing it this way, but I decided to do it nonetheless because we have no other way of checking that the result is not rubbish. Just asserting that the is no exception seems kinda useless. And always just checking if the result table matches will not explicitly test all steps. I believe the SQLPipeline is very closely related to the single steps, e.g. translation, so I think it is acceptable. Maybe just see it as a a kind of "integration" test ;)

Reply:
It is unclean and every time we allow a bit of things like these to trickle into Hyrise we're adding to the pile of poo that will get this project to sink one day.

Reply:
If there is no meaningful way to test what sets `get_optimized_logical_plan()` apart from the functionality that it wraps it is better to leave it untested instead of creating a lie to keep the coverage up (and thus making the coverage a lie as well). I am not saying there IS no meaningful way to test whatever `get_optimized_logical_plan()` is doing, just arguing that IF there is no way, we shouldn't lie about covering its functionality.